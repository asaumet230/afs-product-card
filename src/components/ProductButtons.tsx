import React from 'react';
import { CSSProperties, useCallback, useContext } from 'react';

// Context:
import { ProductContext } from "../context/productContext";

// CSS:
import styles from '../styles/styles.module.css';



export interface Props {
    className?: string;
    style?: CSSProperties;
}

export const ProductButtons = ({ className, style }: Props) => {


    const { counter, increaseBy, maxCount } = useContext(ProductContext);

    const isMaxReached = useCallback(() => (!!maxCount && maxCount === counter), [maxCount, counter]);

    return (
        <div
            style={style}
            className={`${styles.buttonsContainer} ${className}`}>
            <button
                className={styles.buttonMinus}
                onClick={() => increaseBy(-1)}>
                -
            </button>

            <div className={styles.countLabel}>
                {counter}
            </div>

            <button
                className={`${isMaxReached() && styles.disabled} ${styles.buttonAdd}`}
                onClick={() => increaseBy(1)}
                disabled={isMaxReached()}
            >
                +
            </button>
        </div>
    )
}