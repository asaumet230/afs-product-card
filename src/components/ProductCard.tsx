import React from 'react';
import { CSSProperties } from 'react';

// Context:
import { ProductContext } from '../context/productContext';

// Hooks:
import { useProduct } from '../hooks/useProduct';

// CSS:
import styles from '../styles/styles.module.css';

// Interfaces:
import { InitialValues, OnChangeArgs, Product, ProductCartHandlers } from '../interfaces/interfaces';



export interface Props {
    product: Product;
    // children?: ReactElement | ReactElement[];
    children: (args: ProductCartHandlers) => JSX.Element
    className?: string;
    style?: CSSProperties;
    onChange?: (args: OnChangeArgs) => void;
    value?: number;
    initialValues?: InitialValues;
}



export const ProductCard = ({ product, children, className, style, onChange, value, initialValues }: Props) => {

    const { Provider } = ProductContext;
    const { counter, increaseBy, reset, maxCount, isMaxCountReached } = useProduct({
        onChange,
        product,
        value,
        initialValues
    });


    return (

        <Provider value={{
            counter,
            product,
            maxCount,
            increaseBy,
        }}>
            <div
                style={style}
                className={`${styles.productCard} ${className}`}>
                {children({
                    product,
                    increaseBy,
                    reset,
                    maxCount,
                    count: counter,
                    isMaxCountReached,
                })}
            </div>
        </Provider>


    )
}

