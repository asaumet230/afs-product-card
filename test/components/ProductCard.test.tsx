import React from "react";
import { act, create } from "react-test-renderer";

// Component:
import { ProductCard } from '../../src/components';


// Product:
import { product1 } from '../data/products';

describe('Pruebas en el componente <ProductCard />', () => {

    test('Debe de renderizarse correctamente', () => {

        const wrapper = create(
            <ProductCard product={product1}>
                {
                    () => (
                        <h1>Product Card</h1>
                    )
                }
            </ProductCard>
        );

        expect(wrapper.toJSON()).toMatchSnapshot();
    });


    test('Debe Incrementar en 1 el contador', () => {

        const wrapper = create(
            <ProductCard product={product1}>
                {
                    ({ count, increaseBy }) => (
                        <>
                            <h1>Product Card</h1>
                            <span>{count}</span>
                            <button onClick={() => increaseBy(1)}>Incrementar</button>
                        </>
                    )
                }
            </ProductCard>
        );

        let tree = wrapper.toJSON();
        expect(tree).toMatchSnapshot();


        act(() => {
            (tree as any).children[2].props.onClick(); /* fire events that update state */
        });


        tree = wrapper.toJSON();
        expect((tree as any).children[1].children[0]).toBe('1');

    });


});