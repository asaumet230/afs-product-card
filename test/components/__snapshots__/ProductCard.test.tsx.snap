// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <ProductCard /> Debe Incrementar en 1 el contador 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Product Card
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  >
    Incrementar
  </button>
</div>
`;

exports[`Pruebas en el componente <ProductCard /> Debe de renderizarse correctamente 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Product Card
  </h1>
</div>
`;
